### PR TITLE
add http2 health check for SetTransportDefaults

### LINKED ISSUE #374: Client should expose a mechanism to close underlying TCP connections
Re kubernetes/kubernetes#52176, kubernetes/kubernetes#56720, #342 

(Short form: a stalled TCP connection to apiserver from kubelet or kube-proxy can cause ~15 minutes of disruption across a substantial number of nodes until the local kernel closes the socket.)

I believe we need a mechanism for `requestCanceler.CancelRequest` to invoke `transport.CloseIdleConnections` based on config. It looks like we could do this with a small `http.RoundTripper` built to purpose. I hope to submit a PR with this change shortly. 

I'm not sure if that behavior should be activated on a case-by-case basis using `config.WrapTransport` (less invasive, narrower change) or if it should be part of the core config used in `transport.HTTPWrappersForConfig` (needed in multiple use cases per issues listed above). What's the convention here?

### PR DESCRIPTION
**Pull request by duyanghao**
Signed-off-by: duyanghao <1294057873@qq.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

Fixes [Client should expose a mechanism to close underlying TCP connections](https://github.com/kubernetes/client-go/issues/374).

This PR should be merged after [add http2 health check parameters for ConfigureTransport](https://github.com/golang/net/pull/84) merged.

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 1]
```diff
@@ -132,7 +133,7 @@ func SetTransportDefaults(t *http.Transport) *http.Transport {
 	if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
 		klog.Infof("HTTP2 has been explicitly disabled")
 	} else if allowsHTTP2(t) {
-		if err := http2.ConfigureTransport(t); err != nil {
+		if err := http2.ConfigureTransport(t, http2.WithReadIdleTimeout(3*time.Second), http2.WithPingTimeout(2*time.Second)); err != nil {
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:136]: This default value is too frequent, i think 30s is appropriate

Comment (duyanghao) [reply staging/src/k8s.io/apimachinery/pkg/util/net/http.go:136]: I have updated ReadIdleTimeout(30s) and PingTimeout(1s)

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/net/http.go]: 30s might be too infrequent, but we can tune it later.

**Comment by duyanghao** [comment]
@lavalamp This PR requires [add http2 health check parameters for ConfigureTransport](https://github.com/golang/net/pull/84) merged

**Comment by fisherxu** [comment]
Looks like it should be based on this change in golang. https://github.com/golang/net/commit/08b38378de702b893ee869b94b32f833e2933bd2

**Comment by duyanghao** [comment]
> Looks like it should be based on this change in golang. [golang/net@08b3837](https://github.com/golang/net/commit/08b38378de702b893ee869b94b32f833e2933bd2)

Yeah, I have update this PR using the latest http2.ConfigureTransports method.

**Comment by duyanghao** [comment]
do-not-merge/contains-merge-commits

**Comment by aojea** [comment]
@duyanghao have you run `hack/update-vendor.sh`?

**Comment by duyanghao** [comment]
> hack/update-vendor.sh

Not yet, I will have a try.

**Comment by duyanghao** [comment]
> @duyanghao have you run `hack/update-vendor.sh`?

Not yet, I will have a try.

**Comment by aojea** [comment]
@duyanghao can you squash all the commits please?

**Comment by aojea** [comment]
it also fails with the following error
```
Vendor Verify failed.
If you're seeing this locally, run the below command to fix your directories:
hack/update-vendor.sh
These modules are pinned to versions different than the minimal preferred version.
That means that without replace directives, a different version would be selected,
which breaks consumers of our published modules.
1. Use hack/pin-dependency.sh to switch to the preferred version for each module
2. Run hack/update-vendor.sh to rebuild the vendor directory
3. Run hack/lint-dependencies.sh to verify no additional changes are required

golang.org/x/sys
    pinned:    v0.0.0-20200622214017-ed371f2e16b4
    preferred: v0.0.0-20200930185726-fdedc70b468f
    hack/pin-dependency.sh golang.org/x/sys v0.0.0-20200930185726-fdedc70b468f
All pinned versions of checked dependencies match their preferred version.
```
you can check it in the `pull-kubernetes-dependencies` , I think that you can run `make verify` local to verify it

**Comment by povsister** [comment]
I may have opposition to hard-coding those Timeout values.
Internally, there are several ways to detect dead connections. For example: error on `readLoop`, connection idle timeout .. etc.

Most of detections kick-in during the `readLoop` except the one referred here.  The http2 Transport health check will kick-in when the `readLoop` is completely idle for `ReadIdleTimeout` seconds. 
It aims to do an additional health check when there is no incoming traffic for amount of time, but before `IdleConnTimeout` triggers idle connection recycling.
Thus, I think the `ReadIdleTimeout ` should be smaller than `IdleConnTimeout` to achieve it's designed purpose.

In addition, I suggest keep the minimum `PingTimeout` value to default 15s. 
[From the source code in golang.org/x/net/http2/transport.go](https://github.com/golang/net/blob/08b38378de702b893ee869b94b32f833e2933bd2/http2/transport.go#L1806-L1816), It uses `time.AfterFunc()`, not `time.Ticker()`, to trigger health check, and only `Reset()` the timer when the next `readLoop` kicks-in.
It means that the health check will be performed ONLY ONCE when the `readLoop` blocks for amout of time. In such situation, there should be only one PingFrame on the fly, so the `PingTimeout` does not matter much.
But setting the `PingTimeout` too low (eg: 2s in this PR) may lead to incorrect connection recycling during a network jitter or system load spike.

Meanwhile, as the wide acceptance of http2, Kubernetes may have more additional default http2 Transport options to apply in the future. I suggest making a independent and well-commented helper function for the future configuration.

What I described above has already been done in my PR https://github.com/kubernetes/kubernetes/pull/95898
Please consider my solution against this PR.

To be honest(just technically, no offence), this PR looks nasty. It needs rebase&squash, and more comments in the code to describe the purpose of configuration and those "magic number".

**Comment by duyanghao** [comment]
> I may have opposition to hard-coding those Timeout values.
> Internally, there are several ways to detect dead connections. For example: error on `readLoop`, connection idle timeout .. etc.
> 
> Most of detections kick-in during the `readLoop` except the one referred here. The http2 Transport health check will kick-in when the `readLoop` is completely idle for `ReadIdleTimeout` seconds.
> It aims to do an additional health check when there is no incoming traffic for amount of time, but before `IdleConnTimeout` triggers idle connection recycling.
> Thus, I think the `ReadIdleTimeout ` should be smaller than `IdleConnTimeout` to achieve it's designed purpose.
> 
> In addition, I suggest keep the minimum `PingTimeout` value to default 15s.
> [From the source code in golang.org/x/net/http2/transport.go](https://github.com/golang/net/blob/08b38378de702b893ee869b94b32f833e2933bd2/http2/transport.go#L1806-L1816), It uses `time.AfterFunc()`, not `time.Ticker()`, to trigger health check, and only `Reset()` the timer when the next `readLoop` kicks-in.
> It means that the health check will be performed ONLY ONCE when the `readLoop` blocks for amout of time. In such situation, there should be only one PingFrame on the fly, so the `PingTimeout` does not matter much.
> But setting the `PingTimeout` too low (eg: 2s in this PR) may lead to incorrect connection recycling during a network jitter or system load spike.
> 
> Meanwhile, as the wide acceptance of http2, Kubernetes may have more additional default http2 Transport options to apply in the future. I suggest making a independent and well-commented helper function for the future configuration.
> 
> What I described above has already been done in my PR #95898
> Please consider my solution against this PR.
> 
> To be honest(just technically, no offence), this PR looks nasty. It needs rebase&squash, and more comments in the code to describe the purpose of configuration and those "magic number".

@povsister I'm open to any reasonable Suggestions, and I also want to share my opinions:
First of all, I do think users should not care much about the http2 health check, and it's not something Kubernetes even want users to notice, therefore I still suggest to set it by some reasonable default values. Besides, in your ‘proud’ PR, the http2 Transport `ReadIdleTimeout` is configured as the half of http Transport `IdleConnTimeout`, Is this reasonable and not nasty?
Secondly, there's never a 'law' against the PR which needs rebase and squash.

**Comment by JensErat** [comment]
Just adding a [reference to my comments on #95898)](https://github.com/kubernetes/kubernetes/pull/95898#pullrequestreview-518105901), which in the end is discussing both approaches.

Jens Erat <jens.erat@daimler.com>, Daimler TSS GmbH, [Imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

**Comment by duyanghao** [comment]
@JensErat @aojea Since there is no need for uses to configure http2 health check, I do think we should add http2 health check by some reasonable default values, and if you guys think pingTimeout=2s is two short, we can tune it larger, maybe default 15s is a good option.
Besides I can't find any significance that setting http2 Transport ReadIdleTimeout as the half of http Transport IdleConnTimeout, Is there any relationship between these two configurations? 
since PR #95898 looks pretty but unnecessary, again, I insist my way to fix this problem, which is no doubt the most simple and efficient approach to this problem

**Comment by liggitt** [comment]
See https://github.com/kubernetes/kubernetes/pull/95898#issuecomment-717934741 for context on locking.

@caesarxuchao, can you take a look at this and #95898 and give feedback on the best way to configure the PingTimeout/ReadIdleTimeout options?

**Comment by lavalamp** [comment]
This isn't how we have technical disagreements. No amount of technical correctness can make up for treating other people poorly; please don't repeat that behavior. I will close both of these PRs. I'm available via slack or email if more explanation is needed. (additional note: please seek assistance rather than escalating.)

Instead, I've asked @caesarxuchao to send a fix for the issue in question.

**Pull request closed without merging**